src/Hakyll/Core/Compiler/Internal.hs

```diff
@@ -164,25 +165,28 @@ instance MonadMetadata Compiler where
 --------------------------------------------------------------------------------
 instance MonadError [String] Compiler where
   throwError = compilerThrow
-  catchError = compilerCatch
+  catchError = (. matchErr) . compilerCatch
+    where
+      matchErr f Logger.Error es = f es
+      matchErr f _            _  = f []
 
 
 --------------------------------------------------------------------------------
 runCompiler :: Compiler a -> CompilerRead -> IO (CompilerResult a)
 runCompiler compiler read' = handle handler $ unCompiler compiler read'
   where
     handler :: SomeException -> IO (CompilerResult a)
-    handler e = return $ CompilerError [show e]
+    handler e = return $ CompilerError Logger.Error [show e]
 
 
 --------------------------------------------------------------------------------
 instance Alternative Compiler where
-    empty   = compilerThrow []
-    x <|> y = compilerCatch x $ \es -> do
-        logger <- compilerLogger <$> compilerAsk
-        forM_ es $ \e -> compilerUnsafeIO $ Logger.debug logger $
-            "Hakyll.Core.Compiler.Internal: Alternative failed: " ++ e
-        compilerCatch y $ compilerThrow . (es++)
+    empty   = compilerError Logger.Debug []
+    x <|> y = compilerCatch x $ \vx es -> compilerCatch y $ \vy ->
+        case vx `compare` vy of
+            LT -> compilerError vx . const es
+            EQ -> compilerError vx . (es++)
+            GT -> compilerError vy
```

Comment:
I suppose that when dropping a list of messagse because there are more important ones one could log them to the debug screen. Not sure whether necessary.